dialogue-core/src/main/java/com/palantir/dialogue/core/ContentEncodingChannel.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.google.common.net.HttpHeaders;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.EndpointChannel;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.RequestBody;
+import com.palantir.dialogue.Response;
+import com.palantir.logsafe.Preconditions;
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.zip.Deflater;
+import java.util.zip.GZIPOutputStream;
+
+/**
+ * Adds support for transparently encoding sending <code>Content-Encoding: gzip</code> requests
+ * in a client agnostic way based on a Conjure endpoint tag. This requires prior knowledge that
+ * the remote server is capable of handling compressed data.
+ */
+final class ContentEncodingChannel implements EndpointChannel {
+
+    private static final String ENABLEMENT_TAG = "compress-request";
+    private static final String GZIP = "gzip";
+    private static final int BUFFER_SIZE = 8 * 1024;
```

Comment:
For my own learning, how did you pick this buffer size? Just eyeballing something that looks reasonable or are there usual limits we have to account for?

Reply:
8k is fairly standard, it matches guava ByteStreams.copy and (iirc) apache commons IoUtils.copy[Large]. It tends to perform well where larger values don't have much impact on performance. In our webservers we use 16k reusable buffers, however there's limited cost for the large size since they're pooled, yet doesn't impact performance in many cases.
I've done benchmarking, but I can't find my writeup/results at the moment.

Reply:
8k is a common i/o buffer size, used by [guava](https://github.com/google/guava/blob/acb004c10193c1c476a9a9ba8683eeff910e4a82/guava/src/com/google/common/io/ByteStreams.java#L60) and [apache commons-io](https://github.com/apache/commons-io/blob/495abe13df04c96eed0d690202d1c0e053bef29a/src/main/java/org/apache/commons/io/IOUtils.java#L128-L131).
I've done some benchmarking to ensure this is accurate elsewhere, but I can't seem to find my old results.
In this instance, the important piece is that we configure the gzip stream and buffering stream with similar sizes to avoid unnecessary churn, and buffering inputs to the gzip stream such that the cost isn't dominated by interactions with native code (`Deflater` implementation is native, non-java).